src/libsystemd/sd-device/device-enumerator.c

```diff
@@ -599,7 +626,10 @@ static int test_matches(
         if (r <= 0)
                 return r;
 
-        if (!match_property(enumerator, device))
+        if (!match_property(enumerator->match_property, device, /* match_all = */ false))
```

Comment:
Is this correct? Intuitively, I'd expect that all properties are still treated as a single set, just some of them may be required and other optional. E.g., if a device has a property `X` and I match for required property `X` + optional property `Y`, the match should succeed, but currently it would fail (unless I'm missing something).

I think the code should be:
```c
        if (!match_property(enumerator->match_property_required, device, /* match_all = */ true))
                return false;

        if (hashmap_isempty(enumerator->match_property_required)
                if (!match_property(enumerator->match_property, device, /* match_all = */ false))
                        return false;
```

I.e., if all required properties match, it already means that properties match and there's no reason to test the optional properties too. Conversely, if any required property fails to match, it means the properties fail to match and, again, the optional properties are irrelevant. Therefore, the optional properties only need to be tested if there are no required properties.